Observer can cause fatal server error by simply ignoring the socket close

How to reproduce:
- Connect an executor to a session
- Connect an observer to the session
- Pause the application that holds the observer (breakpoint or synchronous `sleep`)
- Disconnect the executor from the session (session closes)
- Create a new session with the same name
- Unpause the observer and send a message

Expected result:
- The observer receives an error
- The application continues work

Actual result:
```
2020-09-19 20:51:20 [notice] server: Pattern recognition server is listening on port 3000
2020-09-19 20:51:27 [info] Task 1: Executor First created
2020-09-19 20:51:27 [info] Task Server: Open new session 'teacher'
2020-09-19 20:51:29 [debug] Observer: Send 'Server: New observer connected'
2020-09-19 20:51:29 [info] Task Server: Connect to teacher session
2020-09-19 20:51:30 [info] Task 1: Close socket
2020-09-19 20:51:30 [info] Task Server: Close the session 'teacher'
2020-09-19 20:51:32 [info] Task 1: Executor First created
2020-09-19 20:51:32 [info] Task Server: Open new session 'teacher'
2020-09-19 20:51:33 [debug] Observer: Receive message {"data": {"message": "Let's start"}}
2020-09-19 20:51:33 [debug] Observer: Unexpected message from observer: '{"data": {"message": "Let's start"}}'
2020-09-19 20:51:33 [info] Observer: Close socket
.../server/WebSocketPool.js:46
      throw new Error(
      ^

Error: The pool doesn't have the socket you want to remove.
    at WebSocketPool.remove (.../server/WebSocketPool.js:46:13)
    at WSTaskServer.removeObserver (.../server/WSTaskServer.js:215:23)
    at WSObserver.observer.afterClose (.../server/WSTaskServer.js:176:14)
    at WSObserver._onClose (.../server/WSClientListener.js:121:10)
    at WebSocket.<anonymous> (.../server/WSClientListener.js:92:35)
    at WebSocket.emit (events.js:327:22)
    at WebSocket.emitClose .../node_modules/ws/lib/websocket.js:191:10)
    at Socket.socketOnClose (.../node_modules/ws/lib/websocket.js:858:15)
    at Socket.emit (events.js:315:20)
    at TCP.<anonymous> (net.js:671:12)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! pattern-recognition-server@0.1.1 start: `node src/server/index.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the pattern-recognition-server@0.1.1 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!    .../.npm/_logs/2020-09-19T17_51_33_926Z-debug.log
```

High-level cause:
- when we disconnect a client, we only check whether the session id it was connected to exists
- we try to remove the client's connection from the pool, but the client wasn't even there for this connection
- the application has an uncaught synchronous error

Low-level cause:
- we use graceful `close` of sockets
- usage of `terminate` solves the problem, but it looks like a workaround https://stackoverflow.com/a/49791634/1305036

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -18,6 +18,11 @@ Added
 
 - Server-side support of JSON messages.
 
+Fixed
+-----
+
+- Observer should not cause server failure by ignoring the socket close.
+
 `0.1.1`_ - 2020-08-23
 =====================
 

--- a/src/server/WSClientListener.js
+++ b/src/server/WSClientListener.js
@@ -83,8 +83,15 @@ class WSClientListener {
   validateSchema() {
   }
 
+  /**
+   * Terminate socket in order to avoid an error
+   * when a client ignores the close request
+   * and sends a message.
+   * The server may try to remove it from the pool
+   * but this causes an error.
+   */
   close() {
-    this.socket.close();
+    this.socket.terminate();
   }
 
   _registerListeners(socket) {

